Issue#79: Implement privacy provider to cover new DB tables

jasonliancatalyst opened the pull request:
This is for issue#79: https://github.com/catalyst/moodle-auth_enrolkey/issues/79

jasonliancatalyst committed:
Issue#79: this commit fixes #79

andrewmadden commented on classes/privacy/provider.php:
```diff
@@ -25,19 +25,240 @@
+        }
+
+        // If current context is course, all users are contained within, get all users.
+        if ($context->contextlevel == CONTEXT_USER) {
```
I believe this is duplication of the check. The early return will catch this.

```
        if (!$context instanceof \context_user) {
            return;
        }
```

andrewmadden commented on classes/privacy/provider.php:
```diff
@@ -25,19 +25,240 @@
+            ];
+
+            $sql = "SELECT usermodified AS userid FROM {auth_enrolkey_redirect} aer
+                        JOIN {enrol} e
```
The enrol table doesn't have anything to do with this plugin. This SQL needs to be looked at.

nhoobin committed:
Issue#79: fixup github workflow actions

This is a test with a step for nodejs. Which may already be installed due to moodle-plugin-ci VendorInstaller.php

andrewmadden commented on classes/privacy/provider.php:
```diff
@@ -25,19 +25,240 @@
+    public static function get_contexts_for_userid(int $userid): contextlist {
+        $sql = "SELECT ctx.id
+                  FROM {user_enrolments} ue
+                  JOIN {enrol} e
```
I don't believe we need the enrol table as this is not an enrol plugin.

Maybe use something like 
```
        $sql = "SELECT ctx.id
                  FROM {auth_enrolkey_redirect} aer
                  JOIN {context} ctx ON ctx.instanceid = aer.usermodified AND ctx.contextlevel = :contextlevel
                 WHERE aer.usermodified = :userid";
        $params = ['userid' => $userid, 'contextlevel' => CONTEXT_USER];
        $contextlist = new contextlist();
        $contextlist->add_from_sql($sql, $params);

        return $contextlist;
```

andrewmadden commented on classes/privacy/provider.php:
```diff
@@ -25,19 +25,240 @@
+    protected static function delete_user_data(int $userid) {
+        global $DB;
+
+        // Because we only use user contexts the instance ID is the user ID.
```
I don't think this comment makes sense, but logic of this method looks good.

andrewmadden commented on classes/privacy/provider.php:
```diff
@@ -25,19 +25,240 @@
+     *
+     * @param  approved_contextlist $contextlist The list of approved contexts for a user.
+     */
+    public static function delete_data_for_user(approved_contextlist $contextlist) {
```
This looks good

nhoobin committed:
Issue#79: resolve github actions phpdoc error

nhoobin committed:
Issue#79: resolve github actions phpunit test error

jasonliancatalyst committed:
Update the context SQL queries

jasonliancatalyst committed:
Merge branch 'issue79_add_privacy_provider' of github.com:catalyst/moodle-auth_enrolkey into issue79_add_privacy_provider

andrewmadden commented:
It looks good to me.
